File: team/vscode-extension/src/project.ts
import * as vscode from 'vscode';
import { getNumberOfProjectFiles, dirSlash, getProjectName } from './utility';
import { getInput } from './input';
import * as fs from 'fs';

// workspacePath has no / at the end
const createNewProject = (projectName: string, workspacePath: string) => {
    if (!getNumberOfProjectFiles(workspacePath)) {
        createNewModule(projectName, workspacePath, projectName);
    }
    else {
        vscode.window.showErrorMessage("This workspace already contains an existing project.");
    }
};

const getModuleTemplate = (moduleName: string) => {
    return `module ${moduleName}(\n);\nendmodule`;
}

const createNewModule = (moduleName: string, workspacePath: string, projectName: string) => {
    const numberOfProjectFiles = getNumberOfProjectFiles(workspacePath);
    if (numberOfProjectFiles > 1) {
        vscode.window.showErrorMessage("More than one project file exists.");
    }
    else {
        const moduleFileName = `${moduleName}.v`;

        // add to .vproj
        try {
            const vprojFilePath = `${workspacePath}${dirSlash}${projectName}.vproj`;
            const vprojLines = fs.existsSync(vprojFilePath) ?
                fs.readFileSync(vprojFilePath, 'utf-8')
                    .split('\n')
                    .filter(Boolean)
                :
                [];
            const lastLine = vprojLines.pop();
            vprojLines.push(moduleFileName);
            if (lastLine) { vprojLines.push(lastLine); }
            const vprojContent = vprojLines.join('\n');
            fs.writeFileSync(vprojFilePath, vprojContent);

        }
        catch (err) {
            vscode.window.showErrorMessage(err.message);
        }

        // create the new module
        try {
            const moduleTemplateContent = getModuleTemplate(moduleName);
            const moduleFilePath = `${workspacePath}${dirSlash}${moduleName}.v`;
            fs.writeFileSync(moduleFilePath, moduleTemplateContent);
        }
        catch (err) {
            vscode.window.showErrorMessage(err.message);
        }

    }
};

export const newProjectHandler = () => {
    const folders = vscode.workspace.workspaceFolders || [];
    if (folders.length === 1) {
        const workspacePath = folders[0].uri.fsPath;
        getInput(`Enter project name`,
            (text: string) => { if (/[a-zA-Z][a-zA-Z0-9_]*/.test(text) && text.length) { return undefined; } return `Enter a valid project name`; })
            .then((projectName) => {
                if (projectName) {
                    createNewProject(projectName, workspacePath);
                }
                else {
                    vscode.window.showErrorMessage(`Enter a valid project name.`);
                }
            });
    }
    else {
        vscode.window.showErrorMessage(`Verishot only supports 1 project folder.`);
    }
};

export const newModuleHandler = () => {
    const folders = vscode.workspace.workspaceFolders || [];
    if (folders.length === 1) {
        const workspacePath = folders[0].uri.fsPath;
        const projectName = getProjectName(workspacePath);
        if (projectName) {
            getInput(`Enter module name`,
                (text: string) => { if (/[a-zA-Z][a-zA-Z0-9_]*/.test(text)) { return undefined; } return `Enter a valid module name`; })
                .then((moduleName) => {
                    if (moduleName && projectName) {
                        createNewModule(moduleName, workspacePath, projectName);
                    }
                    else {
                        vscode.window.showErrorMessage(`Enter a valid module name`);
                    }
                });
        }
        else {
            if (!getNumberOfProjectFiles(workspacePath)) {
                vscode.window.showErrorMessage(`Create a new project first`);
            }
            else {
                vscode.window.showErrorMessage(`More than 1 project files found`);
            }
        }
    }
    else {
        vscode.window.showErrorMessage(`Verishot only supports 1 project folder.`);
    }
};

File: team/Verishot/src/FrontEnd/Project.fs
(*
    Front end code for `Project` managing
    =====================================
    Author: lhl2617
*)


module Verishot.Project

open System.IO
open System.Text.RegularExpressions

open Verishot.FrontEnd
open Verishot.Util
open Verishot.FileUtil
open Verishot.FrontEndParser

let getNumberOfVProjFiles workspacePath = 
    workspacePath
    |> getFilenamesInFolderPath "*.vproj"
    |> List.length

let getModuleTemplate moduleName = 
    sprintf "module %s (\n);\nendmodule" moduleName

let createNewModule vProjFilePath moduleName = 
    let exitCode = exitCodes.NewModuleError
    match matchIdentifierRegex moduleName with
    | true -> 
        let currentModules = getExistingModuleNamesFromVProj vProjFilePath
        match List.exists ((=) moduleName) currentModules with 
        | false ->
            // add to .vproj
            let vProjContent = readFileToString vProjFilePath + sprintf "\n%s.v" moduleName
            writeStringToFile vProjFilePath vProjContent

            // create the new module
            let workspacePath = Directory.GetParent(vProjFilePath).FullName
            let moduleTemplateContent = getModuleTemplate moduleName
            let moduleFilepath = workspacePath +/ moduleName + ".v"
            writeStringToFile moduleFilepath moduleTemplateContent

            let stdout = sprintf "Successfully created module `%s`" moduleName
            Ok stdout
        | _ -> 
            let stdout = "Module name already exists"
            Error (exitCode, stdout)
    | _ -> 
        let stdout = "Enter a valid module name"
        Error (exitCode, stdout)

let deleteModule vProjFilePath moduleName = 
    let exitCode = exitCodes.DeleteModuleError
    let currentModules = getExistingModuleNamesFromVProj vProjFilePath
    match List.exists ((=) moduleName) currentModules with 
    | true ->
        // update .vproj
        let vProjContent =
            vProjFilePath
            |> readFileToStringList
            |> List.filter (fun x -> x.Trim() |> (<>) (moduleName + ".v"))
            |> String.concat "\n"
        writeStringToFile vProjFilePath vProjContent

        // delete .v
        let workspacePath = Directory.GetParent(vProjFilePath).FullName
        let vFilePath = workspacePath +/ moduleName + ".v"
        deleteFile vFilePath

        let stdout = sprintf "Successfully deleted module `%s`" moduleName
        Ok stdout
    | _ -> 
        let stdout = sprintf "Module `%s` does not exist" moduleName
        Error (exitCode, stdout)


let createNewProject workspacePath projectName = 
    let exitCode = exitCodes.NewProjectError
    
    let vProjHeader = "// ===== Verishot Project File =====
// WARNING: Do not edit this file directly unless you know what you are doing!
// 
// The first module listed must be the top level module"

    match getNumberOfVProjFiles workspacePath with 
    | 0 -> 
        match matchIdentifierRegex projectName with
        | true -> 
            let vProjFilePath = workspacePath +/ projectName + ".vproj"
            writeStringToFile vProjFilePath vProjHeader

            match createNewModule vProjFilePath projectName with
            | Ok stdout ->
                let stdout2 = sprintf "Successfully created project `%s`" projectName
                let stdout' = stdout +@ stdout2
                Ok stdout'
            | Error x -> Error x           
        | _ ->
            let stdout = sprintf "Enter a valid project name"
            Error (exitCode, stdout)
    | _ -> 
        let stdout = sprintf "This workspace already contains an existing project"
        Error (exitCode, stdout)

let listModules vProjFilePath = 
    getExistingModuleNamesFromVProj vProjFilePath 
    |> String.concat "\n" 
    |> Ok